### ISSUE TITLE
Updating Calico-Kube-Controller from 3.10 to 3.11 inadvertidely caused deletion of all nodes and workloadendpoints 

### ISSUE THREAD

**Issue by mrrandrade** [projectcalico/calico#3187]
At the end of our main Production cluster, as the last component, we updated Calico-Kube-Controller from 3.10.2 to 3.11.2.

For some unknown reason, all entries for calico nodes and workloadendpoints got removed from our ETCD database.

As side effects, all of our 5000+ pods lost connectivity and had to be restarted.

A downgrade from 3.11 to 3.10 also triggered the event - all nodes and workloadendpoints got removed again.

A valid sidenote: we've updated 4 clusters, and it only happened on 1 (the important one, of course), so it most likely irreproduceable.

## Expected Behavior
 Flawless update and beers;

## Current Behavior
 Fire and brimstone and long hours to recover everything.

## Possible Solution

I wouldn't say a solution, but we removed 'node' from enabled controllers for the kube_controller since this misbehaviour:

```
    - name: ENABLED_CONTROLLERS
      value: policy,namespace,serviceaccount,workloadendpoint
```

## Steps to Reproduce (for bugs)

Its probably irreproduceable, since it didn't happen before on our 'trial' updates.

## Context
 

I'd guess that it might be due to stale data on our ETCD datastores.

We use IP addresses of our nodes as their nodename, but neither IPs or names changed during migration.

It happened both ways: 3.10 to 3.11, and rolling back from 3.11 to 3.10.

It was also triggered before, after we inserted some 3.11 calico-node machines on a mainly 3.10 cluster.

Even though it might be impossible to pinpoin the cause, since this is a soul crushing and major breaking event, if you guys want to, I'd accept some suggestions of scenarios to try and identify what happened. I.e., I wonder if maybe if I spin up some VMs that roll ETCD backups and 1 kubernetes master running we might map what caused it.

## Your Environment

* Calico version: 3.11.2 with ETCD datastore;
* Orchestrator version (e.g. kubernetes, mesos, rkt): Kubernetes, kube controller running as pod;
* Operating System and version: CoreOS 2303.3.0
* Link to your project (optional):

**Comment by caseydavenport** [comment]
@mrrandrade yikes, I'm sorry this happened to you.

Do you happen to have logs from the kube-controllers pod from when this occurred? It might contain information about what triggered this behavior. 

I'd really like to understand what happened here and would love to work with you to identify the cause. A few things come to mind:
- What are the differences between the successful trial environments and the prod env? Are the nodes named in the same way? Are they running the same configuration? Anything else?
- Do you use any modifications to the Calico manifests? 
- Was this just a Calico upgrade? Or was anything else changing in the cluster at the same time (e.g., k8s version or config?)

kube-controllers will clean up nodes and weps on that node if it thinks the node no longer exists in Kubernetes. It does this using a field on the Calico `Node` object which includes a reference to the name of the node in the Kubernetes API (`node.Spec.OrchestratorReferences`). 

If the controller sees that an orchestrator reference to the kubernetes node exists, but no corresponding node exists in the Kubernetes API, it will assume the node has been removed and clean up data related to that node. Some things to check, if you are able to reproduce this, would be:

- Do the `spec.OrchRef` values in `calicoctl get nodes -o yaml` change from before the upgrade / downgrade to after?
- Do the names of the nodes in Kubernetes change at all, even momentarily?

**Comment by mrrandrade** [comment]
> What are the differences between the successful trial environments and the prod env? 

Basically, one is massively used (5000+ pods) and the others are barely used (some not at all, the others with 100+ pods); we have no sense of proportions :-P

> Are the nodes named in the same way? 

Back in the days, nodes were named using their ip addressess, and we kept that way.

```
apiVersion: projectcalico.org/v3
kind: Node
metadata:
  creationTimestamp: "2020-01-26T02:12:09Z"
  name: 10.69.40.69
  resourceVersion: "218813355"
  uid: dbda5c3a-08e1-4a26-ab72-5f6763ce831c
spec:
  bgp:
    ipv4Address: 10.69.40.69/32
  orchRefs:
  - nodeName: 10.69.40.69
    orchestrator: k8s

```
> Are they running the same configuration? 

Pretty much, althought, for this version, we changed from using Calico Node as a unit on systemd to the calico-node daemonset. 

> Do you use any modifications to the Calico manifests?

Well, we basically had systemd units to cover calico-node and cni, so I'm guessing that's a yes :-P

The calico-kube-controller manifest has no changes other than some tolerations and node Selector.

Calico-node has a lot more customizations due to bird.cfg template change, but I'm not sure if any of them would affect anything. 

We did change this so it registers itself with the nodeIP:

```
        - name: NODENAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: status.podIP
```

> Was this just a Calico upgrade? Or was anything else changing in the cluster at the same time (e.g., k8s version or config?)

Its the 'change day' that happens every 2 months, so we change basically everything. For instance:
- Kubernetes 1.16.3 to 1.17.2
- CoreOS from 2247.7.0 to 2303.3.0

We changed a lot more stuff (coredns, ingressess) but I believe those up there are the only relevant ones.

> Anything else?

Maybe the most relevant thing is we actually migrated from systemd units to calico 'official' manifests for the Kubernetes cluster. That means now we're using the initContainer for calico-cni.

We did change the cniVersion to 0.3.1 from 0.3.0; this is the 10-calico-conflist for our setup:

```
{
  "name": "k8s-pod-network",
  "cniVersion": "0.3.1",
  "plugins": [
    {
      "type": "calico",
      "etcd_endpoints": "https://etcd1:2379,https://etcd2:2379,https://etcd3:2379",
      "etcd_key_file": "/etc/cni/net.d/calico-tls/tls.key",
      "etcd_cert_file": "/etc/cni/net.d/calico-tls/tls.crt",
      "etcd_ca_cert_file": "/etc/cni/net.d/calico-tls/tls.ca",
      "log_level": "info",
      "mtu": 1440,
      "ipam": {
         "ipv4_pools": ["192.168.0.0/18"],
         "type": "calico-ipam"
        },
      "policy": {
         "type": "k8s",
         "k8s_api_root": "https://kubernetes.apiserver:443",
         "k8s_auth_token": "...token..."
       },
      "kubernetes": {
         "kubeconfig": "/etc/cni/net.d/calico-kubeconfig"
      }
    },
    {
      "type": "portmap",
      "snat": true,
      "capabilities": {"portMappings": true}
    }
  ]
}
```

There's no entry for nodename on the CNI, but it just works.

**Comment by caseydavenport** [comment]
Thanks for the info! Do you happen to have a section in your new `calico.yaml` that looks like this?

```
            # Set noderef for node controller.
            - name: CALICO_K8S_NODE_REF
              valueFrom:
                fieldRef:
                  fieldPath: spec.nodeName
```

That's the field that is actually used for correlating the Kubernetes node with the Calico one. by default that uses the spec.nodeName, which I would assume is the same as the node IP and might just work, but if for whatever reason this was set to something different it could cause symptoms like the one you're describing. 

I'm guessing based on the node you pasted above that this was properly set in the systemd unit files, so worth checking if it changes in the switch at all. 

> Maybe the most relevant thing is we actually migrated from systemd units to calico 'official' manifests for the Kubernetes cluster.

Yeah, this feels like a likely culprit as well. I doubt it's the init container / CNI config, but it's possible that the environment variable config passed to calico/node is subtly different between the two approaches. Would you be able to share your systemd unit file from before, and the new `calico.yaml` file you're using now?

**Comment by caseydavenport** [comment]
@mrrandrade any updates on this issue? Would really like to get to the bottom of it if we can!

**Comment by rafaelvanoni** [comment]
@mrrandrade Could you give us an update on this one? Thanks

**Comment by spikecurtis** [comment]
Closing this out under the assumption that you all got to the bottom of your issue.  Please re-open if you still need help!

**Comment by marcelo-devsres** [comment]
Hey, I came here to this long forgotten issue to apologize, I got realocated to something else that was of a higher priority so had to just pretend everything was "normal" and move on. I only get about 1 day or 2 in a semester window to figure out stuff about this problem, which clearly is not enough.

Not only that, but I switched github accounts, which means I didn't get the notifications of the follow ups.

Again, truly sorry for "having forgotten" about this and no, no one actually got to the botton of the issue (we still run calico without the node controller because of the fears).
